File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
input = sys.stdin.readline

n = int(input())
arr = [0]
for _ in range(n):
    arr.append(int(input()))

dp = [0] * (n+1)

def sol(n):
    dp[1] = arr[1]
    if n == 1:
        return dp[1]
    dp[2] = arr[1] + arr[2]
    if n == 2:
        return dp[2]
    dp[3] = max(arr[2] + arr[3], arr[1] + arr[3])
    if n == 3:
        return dp[3]
    for i in range(4, n+1):
        dp[i] = arr[i] + max(dp[i-3] + arr[i-1], dp[i-2])
    return dp[-1]

print(sol(n))

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n + 1];
        int[] dp = new int[n + 1];
        for (int i = 1; i < n + 1; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 1; i < n + 1; i++) {
            if (i == 1) {
                dp[i] = arr[i];
            } else if (i == 2) {
                dp[i] = dp[i - 1] + arr[i];
            } else if (i == 3) {
                dp[i] = Math.max(dp[i - 2], arr[i - 1]) + arr[i];
            } else {
                dp[i] = Math.max(dp[i - 2], dp[i - 3] + arr[i - 1]) + arr[i];
            }
        }

        System.out.println(dp[n]);

        br.close();
    }


}

